File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        cash = 0

        def traverse(node, max_v):
            nonlocal cash
            if not node:
                return
            # print(node.val, max_v)

            if max_v <= node.val:
                cash += 1
            
            max_v = max(max_v, node.val)
            traverse(node.right, max_v)
            traverse(node.left, max_v)
        
        traverse(root, root.val-1)
        return cash

File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-1.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func buildTree(preorder []int, inorder []int) *TreeNode {
    
	// parameters are slices so its fine

	inmap := map[int]int{}

    for ind, val := range inorder{
        inmap[val] = ind
    }

    var buildFunc func(int, int) *TreeNode
    pre := 0
    buildFunc = func(li, ri int) *TreeNode{

        if ri < li {
            return nil
        }

        node := &TreeNode { Val: preorder[pre]}
        k := inmap[preorder[pre]]
        pre++

        // left_size := k-li
        node.Left = buildFunc(li, k-1)
        node.Right = buildFunc(k+1, ri)
        return node
    }

    return buildFunc(0, len(inorder)-1)
}
